### PR TITLE
android bump - kotlin/haptic/timeout

### LINKED ISSUE #5374: FR: Add rumble support
More of a brainstorm than an FR. Linkdump without commentary incoming.

Docs:
https://wiki.libsdl.org/SDL_HapticOpenFromJoystick
https://developer.android.com/reference/android/os/Vibrator

https://davidgow.net/handmadepenguin/ch6.html

Background:
https://medium.com/@sdrzn/make-your-ios-app-feel-better-a-comprehensive-guide-over-taptic-engine-and-haptic-feedback-724dec425f10

### PR DESCRIPTION
**Pull request by pazos**
Visible changes:

- KOReader icon on start & friendly e-ink spinner, authored by @ColinKinloch

ezgif-3-83c30eeb53a4

- Safer and more verbose WRITE_SETTINGS permissions: 

ezgif-2-91e834e26ff3

Under the hood: 

- Migration to Kotlin (everything except the test activity and the e-ink methods to refresh the screen):
	+ Safety against null pointer exceptions (It can still be triggered from JNI)
	+ Less boilerplate code.
	+ Usually easier to read and more expresive.
	
- More responsive garbage collector:
	+ avoid leaking activity context
	+ slightly more memory usage on average (~1MB) but with the hability to free some memory on-demand (2-3MB)

- Removed wakelock permissions.

Fixes #5374 
Requires https://github.com/koreader/android-luajit-launcher/pull/193

### CODE REVIEW

**Comment by pazos** [comment]
I'm done with testing on my end. Probably others will find some issues that I will be happy to fix :smile:

File: frontend/ui/elements/screen_android.lua [hunk 1]
```diff
@@ -32,13 +32,35 @@ local function humanReadableTimeout(timeout)
...
     G_reader_settings:saveSetting("android_screen_timeout", timeout)
-    android.setScreenOffTimeout(timeout)
+    android.timeout.set(timeout)
+end
+
+local function requestWriteSettings()
+    local UIManager = require("ui/uimanager")
+    local ConfirmBox = require("ui/widget/confirmbox")
+    UIManager:show(ConfirmBox:new{
+        text = _("Allow KOReader to modify system settings?\n\nYou will be prompted with a permission management screen. You'll need to give permission to KOReader and start the program again"),
```
Comment (Frenzie) [inline frontend/ui/elements/screen_android.lua]: Maybe something like this? (Btw, you missed the period at the end. ;-) )
```suggestion
        text = _("Allow KOReader to modify system settings?\n\nYou will be prompted with a permission management screen. You'll need to give KOReader permission and then restart the program."),
```

Comment (pazos) [reply frontend/ui/elements/screen_android.lua]: thanks!

**Comment by pazos** [comment]
Still unimplemented: Ui menu to override vibration (ie: vibration on KOReader when is disabled on system settings. 

Can wait a bit. First merge this PR and wait for broken reports :smile:

**Comment by poire-z** [comment]
Dunno if it's related to these changes, but I get a crash when I highlight some text and choose "Copy" in the button dialog:
```
10-09 09:09:30.538  8751  8763 E AndroidRuntime: FATAL EXCEPTION: Thread-277
10-09 09:09:30.538  8751  8763 E AndroidRuntime: Process: org.koreader.launcher, PID: 8751
10-09 09:09:30.538  8751  8763 E AndroidRuntime: java.lang.RuntimeException: Can't create handler inside thread that has not called Looper.prepare()
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at android.os.Handler.<init>(Handler.java:200)
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at android.os.Handler.<init>(Handler.java:114)
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at android.content.ClipboardManager$2.<init>(ClipboardManager.java:69)
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at android.content.ClipboardManager.<init>(ClipboardManager.java:69)
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at android.app.SystemServiceRegistry$10.createService(SystemServiceRegistry.java:220)
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at android.app.SystemServiceRegistry$10.createService(SystemServiceRegistry.java:217)
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at android.app.SystemServiceRegistry$CachedServiceFetcher.getService(SystemServiceRegistry.java:806)
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at android.app.SystemServiceRegistry.getSystemService(SystemServiceRegistry.java:759)
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at android.app.ContextImpl.getSystemService(ContextImpl.java:1351)
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at android.content.ContextWrapper.getSystemService(ContextWrapper.java:627)
10-09 09:09:30.538  8751  8763 E AndroidRuntime:        at org.koreader.launcher.BaseActivity.setClipboardText(BaseActivity.kt:223)
10-09 09:09:30.542  1528  2916 D ActivityManager: New dropbox entry: org.koreader.launcher, data_app_crash, aa5db121-a460-4a2b-b665-98c09cacd273
10-09 09:09:30.544  1528  2916 W ActivityManager:   Force finishing activity org.koreader.launcher/.MainActivity
```
(That `New dropbox entry` stuff, it's not us, it must be some Sony or Android stuff, right?)

**Comment by Frenzie** [comment]
Wow, on my Sailfish XA2 the Android system constant is nothing like the native Sailfish keypress haptic feedback. (Probably a minor bug, except since you don't use any Android keyboards one that's unlikely to pop up.) On my Zenfone 2 it's also *extremely* present.

`CONTEXT_CLICK` and `LONG_PRESS` are basically the kind of haptic feedback I normally feel on touch, and the `KEYBOARD_TAP` one isn't used by the Zen keyboard.

The system "Vibrate on touch" setting works correctly to control it, but unfortunately on/off is the only choice you have. Nothing about the type or duration of vibration.

So, working quite nicely :+1: but it might need a touch more thought. :-)

**Comment by pazos** [comment]
@poire-z: thanks for the report. In fact the clipboard was broken:

- on copy for 4.0 to 7.1
- on paste for all devices (I forgot to change the signature of the JNI call).

**Comment by pazos** [comment]
@Frenzie:

> Wow, on my Sailfish XA2 the Android system constant is nothing like the native Sailfish keypress haptic feedback. (Probably a minor bug, except since you don't use any Android keyboards one that's unlikely to pop up.) On my Zenfone 2 it's also extremely present.

> CONTEXT_CLICK and LONG_PRESS are basically the kind of haptic feedback I normally feel on touch, and the KEYBOARD_TAP one isn't used by the Zen keyboard.

Exactly the same on a Moto G5S.

> The system "Vibrate on touch" setting works correctly to control it, but unfortunately on/off is the only choice you have. Nothing about the type or duration of vibration per type of haptic feedback.

Unless you install a custom rom or your vendor has extra settings that's the best we can get using this API.

> So, working quite nicely +1 but it might need a touch more thought. :-)

Vibration api works quite well. It lets us override amplitude/intensity (0-255) and duration (0-1000ms) on all supported apis (new ones have fancy vibration effects but that aside).

Switching to the vibration api is easy but we need to provide our own settings inside the app:
we need to provide a set of contexts and associate each one with a custom intensity and duration andd provide a way to replace default values.

**Comment by Frenzie** [comment]
> Switching to the vibration api is easy but we need to provide our own settings inside the app:
we need to provide a set of contexts and associate each one with a custom intensity and duration andd provide a way to replace default values.

For me personally that concept would work just fine without the Vibration API. I'd just tell it to use something like `LONG_PRESS` instead of `KEYBOARD_TAP`.

**Pull request merged**